src/site/pages/digital/src/containers/ExprToCircuitPopup/index.tsx

```diff
@@ -0,0 +1,201 @@
+    //  so that the components are not literally in the uppermost leftmost corner
+    const startPos = info.camera.getPos().sub(info.camera.getCenter().scale(info.camera.getZoom()/1.5));
+    OrganizeMinDepth(circuit, startPos);
+    const action = new GroupAction([CreateDeselectAllAction(info.selections)]);
+    if (isIC) {
+        const data = ICData.Create(circuit);
+        data.setName(expression);
+        const ic = new IC(data);
+        ic.setName(expression);
+        ic.setPos(info.camera.getPos());
+        action.add(new CreateICDataAction(data, info.designer));
+        action.add(new PlaceAction(info.designer, ic));
+        action.add(new SelectAction(info.selections, ic));
+    }
+    else {
+        action.add(CreateAddGroupAction(info.designer, circuit));
+        action.add(CreateGroupSelectAction(info.selections, circuit.getComponents()));
+    }
+    info.history.add(action.execute());
+    info.renderer.render();
+}
+
+export const ExprToCircuitPopup = (() => {
+    return ({ mainInfo }: Props) => {
+        const {curPopup} = useSharedSelector(
+            state => ({ curPopup: state.header.curPopup })
+        );
+        const dispatch = useSharedDispatch();
+        let defaultUserOps = {} as Record<FormatLabels, string>;
+        for (let [key, value] of FormatMap.get("|")) {
+            defaultUserOps[key] = value;
+        }
+
+        const [expression, setExpression] = useState("");
+        const [errorMessage, setErrorMessage] = useState("");
+        const [isIC, setIsIC] = useState(false);
+        const [input, setInput] = useState("Switch");
+        const [format, setFormat] = useState("|");
+        const [userOps, setUserOps] = useState(defaultUserOps);
+
+        const inputTypes = Array.from(Inputs.keys()).map((input) =>
+            <option key={input} value={input}>{input}</option>
+        );
+    
+        const formats = Array.from(FormatMap.entries()).map((formatEntry) =>
+            <div key={formatEntry[0]}>
+                <input type="radio" id={formatEntry[0]} name="format" checked={format === formatEntry[0]} onChange={() => setFormat(formatEntry[0])} value={formatEntry[0]} />
+                <label htmlFor={formatEntry[0]}>{formatEntry[1].get("label")}</label><br/>
+            </div>
+        );
+
+        const customOps: [string, FormatLabels][] = [
+            ["AND", "&"],
+            ["OR", "|"],
+            ["XOR", "^"],
+            ["NOT", "!"],
+            ["(", "("],
+            [")", ")"],
+            ["Separator", "separator"],
+        ];
+        const customOpsInput = customOps.map((op) => 
+            <div className="exprtocircuit__popup__customOps" key={op[0]}>
+                <input title={"Enter symbol for " + op[0]} type="text" value={userOps[op[1]]} onChange={e => setUserOps({...userOps, [op[1]]: e.target.value})}/>
+                <label htmlFor={userOps[op[1]]}>Custom {op[0]}: "{userOps[op[1]]}"</label>
+            </div>
+        );
```

Comment:
This can be its own React component called like `CustomOps` (same folder, new file `CustomsOps.tsx`)

The state `userOps` can stay here, and the `CustomOps` component can just take whatever it needs as props